### ISSUE TITLE
jax.nn.dot_product_attention ignores local_window_size when mask is not present

### ISSUE THREAD

**Issue by hamzamerzic** [jax-ml/jax#32300]
### Description

Code that demonstrates this (run in a TPU Colab):

```python
import jax
import jax.numpy as jnp

BATCH_SIZE = 32
T_q = 2048
T_kv = 2080
NUM_HEADS = 8
HEAD_DIM = 64

key = jax.random.PRNGKey(42)
q_key, k_key, v_key, m_key = jax.random.split(key, 4)
q = jax.random.normal(
    q_key, (BATCH_SIZE, T_q, NUM_HEADS, HEAD_DIM), dtype=jnp.bfloat16
)
k = jax.random.normal(
    k_key, (BATCH_SIZE, T_kv, NUM_HEADS, HEAD_DIM), dtype=jnp.bfloat16
)
v = jax.random.normal(
    v_key, (BATCH_SIZE, T_kv, NUM_HEADS, HEAD_DIM), dtype=jnp.bfloat16
)

jitted_attention = jax.jit(
    lambda q, k, v, m, l: jax.nn.dot_product_attention(
        q, k, v, mask=m, local_window_size=(0, l - 1)
    ),
    static_argnums=4,
)

print(
    "Attention outputs match for different window sizes (32 vs 1) with no mask:",
    jnp.all(
        jitted_attention(q, k, v, None, 32)
        == jitted_attention(q, k, v, None, 1)
    ),
)

random_mask = jax.random.uniform(m_key, (BATCH_SIZE, 1, T_q, T_kv)) > 0.5

print(
    "Attention outputs match for different window sizes (32 vs 1) with a mask:",
    jnp.all(
        jitted_attention(q, k, v, random_mask, 32)
        == jitted_attention(q, k, v, random_mask, 1)
    ),
)
```

This code results in:
```
Attention outputs match for different window sizes (32 vs 1) with no mask: True
Attention outputs match for different window sizes (32 vs 1) with a mask: False
```

The culprit seems to be [this line](https://github.com/jax-ml/jax/blob/a589b31e20745d7ce7153d8fa8d7f815418d4dd4/jax/_src/nn/functions.py#L896) which prevents the local mask to get built [here](https://github.com/jax-ml/jax/blob/a589b31e20745d7ce7153d8fa8d7f815418d4dd4/jax/_src/nn/functions.py#L910).

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.8.0
jaxlib: 0.8.0
numpy:  2.3.3
python: 3.12.11 (stable, redacted, redacted) [Clang 9999.0.0 (fc44a4fcd3c54be927c15ddd9211aca1501633e7)]
device info: TPU v3-8, 8 local devices"
process_count: 1
platform: uname_result(system='Linux', node='6896a6a54d5a17c6-36316f7d32d.borgtask.google.com', release='5.10.0-smp-1107.69.0.0', version='#1 [v5.10.0-1107.69.0.0] SMP @1754001642', machine='x86_64')
XLA_PYTHON_CLIENT_PREALLOCATE=false
```

**Comment by justinjfu** [comment]
@kaixih Do you have time to take a look?

**Comment by akshay-babbar** [comment]
Hi @justinjfu / @hamzamerzic  -  I'd like to work on this if no one is actively on it yet.

  I believe adding a check for local_window_size to the initial condition in _apply_masks will resolve this.

  If no one is working on this, I would be happy to submit a pull request with the fix and a corresponding test case.

Thanks!

**Comment by kaixih** [comment]
@akshay-babbar Thx. Please. I am tied up with something else rn, please go ahead.
